src/fft_2n.h

```diff
@@ -284,39 +339,136 @@ template <typename T>
 void Radix2<T>::fft(vec::Buffers<T>& output, vec::Buffers<T>& input)
 {
     const unsigned len = this->n;
-    const unsigned size = this->pkt_size;
     const unsigned input_len = input.get_n();
     // to support FFT on input vectors of length greater than from `data_len`
     const unsigned group_len =
         (input_len > data_len) ? len / input_len : len / data_len;
 
+    const std::vector<T*>& i_mem = input.get_mem();
+    const std::vector<T*>& o_mem = output.get_mem();
+
     for (unsigned idx = 0; idx < input_len; ++idx) {
         // set output  = scramble(input), i.e. bit reversal ordering
-        T* a = input.get(idx);
-        const unsigned end = rev[idx] + group_len;
-        for (unsigned i = rev[idx]; i < end; ++i) {
-            output.copy(i, a);
+        for (unsigned i = rev[idx]; i < rev[idx] + group_len; ++i) {
+            memcpy(o_mem[i], i_mem[idx], buf_size);
         }
     }
     for (unsigned idx = input_len; idx < data_len; ++idx) {
         // set output  = scramble(input), i.e. bit reversal ordering
-        const unsigned end = rev[idx] + group_len;
-        for (unsigned i = rev[idx]; i < end; ++i) {
-            output.fill(i, 0);
+        for (unsigned i = rev[idx]; i < rev[idx] + group_len; ++i) {
+            memset(o_mem[i], 0, buf_size);
         }
     }
-    // perform butterfly operations
-    for (unsigned m = group_len; m < len; m *= 2) {
-        const unsigned doubled_m = 2 * m;
+
+    // ----------------------
+    // Two layers at a time
+    // ----------------------
+    unsigned m = group_len;
+    const unsigned end = len / 2;
+    for (; m < end; m <<= 2) {
         for (unsigned j = 0; j < m; ++j) {
-            const T r = W->get(j * len / doubled_m);
-            for (unsigned i = j; i < len; i += doubled_m) {
-                T* a = output.get(i);
-                T* b = output.get(i + m);
+            butterfly_ct_two_layers_step(output, j, m);
+        }
+    }
+    if (m < len) {
+        assert(m == end);
+        // perform the last butterfly operations
+        for (unsigned j = 0; j < m; ++j) {
+            const T r = W->get(j);
+            butterfly_ct_step(output, r, j, m, len);
+        }
+    }
+}
 
-                // perform butterfly operation for Cooley-Tukey FFT algorithm
-                this->gf->butterfly_ct(r, a, b, size);
-            }
+// for each pair (P, Q) = (buf[i], buf[i + m]):
+// P = P + c * Q
+// Q = P - c * Q
+template <typename T>
+void Radix2<T>::butterfly_ct_step(
```

Comment:
The code seems identical to `butterfly_ct_step_offset`, with `offset=0`, so I think the code could be merged by giving offset a default value of `0`.

Or imply get rid of `butterfly_ct_step_offset`, as it seems unused.

Reply:
I aggregated `butterfly_ct_step` and `butterfly_ct_step_offset` :)